fem/lininteg_boundary.cpp

```diff
@@ -0,0 +1,302 @@
+                  }
+               }
+               else
+               {
+                  coeff_val = cst ? C(0,0,0,0) : C(0,x,y,e);
+               }
+               QQ(y,x) = W(x,y) * coeff_val * detJ(x,y,e);
+            }
+         }
+         MFEM_SYNC_THREAD;
+         MFEM_FOREACH_THREAD(qy,y,q)
+         {
+            MFEM_FOREACH_THREAD(dx,x,d)
+            {
+               double u = 0.0;
+               for (int qx = 0; qx < q; ++qx) { u += QQ(qy,qx) * Bt(dx,qx); }
+               QD(qy,dx) = u;
+            }
+         }
+         MFEM_SYNC_THREAD;
+         MFEM_FOREACH_THREAD(dy,y,d)
+         {
+            MFEM_FOREACH_THREAD(dx,x,d)
+            {
+               double u = 0.0;
+               for (int qy = 0; qy < q; ++qy) { u += QD(qy,dx) * Bt(dy,qy); }
+               Y(dx,dy,c,e) += u;
+            }
+         }
+         MFEM_SYNC_THREAD;
+      }
+   });
+}
+
+static void BLFEvalAssemble(const FiniteElementSpace &fes,
+                            const IntegrationRule &ir,
+                            const Array<int> &markers,
+                            const Vector &coeff,
+                            const bool normals,
+                            Vector &y)
+{
+   Mesh &mesh = *fes.GetMesh();
+   const int dim = mesh.Dimension();
+   const FiniteElement &el = *fes.GetBE(0);
+   const MemoryType mt = Device::GetDeviceMemoryType();
+   const DofToQuad &maps = el.GetDofToQuad(ir, DofToQuad::TENSOR);
+   const int d = maps.ndof, q = maps.nqpt;
+   int flags = FaceGeometricFactors::DETERMINANTS;
+   if (normals) { flags |= FaceGeometricFactors::NORMALS; }
+   const FaceGeometricFactors *geom = mesh.GetFaceGeometricFactors(
+                                         ir, flags, FaceType::Boundary, mt);
+   auto ker = (dim == 2) ? BLFEvalAssemble2D<> : BLFEvalAssemble3D<>;
+
+   if (dim==2)
+   {
+      if (d==1 && q==1) { ker=BLFEvalAssemble2D<1,1>; }
+      if (d==2 && q==2) { ker=BLFEvalAssemble2D<2,2>; }
+      if (d==3 && q==3) { ker=BLFEvalAssemble2D<3,3>; }
+      if (d==4 && q==4) { ker=BLFEvalAssemble2D<4,4>; }
+      if (d==5 && q==5) { ker=BLFEvalAssemble2D<5,5>; }
+      if (d==2 && q==3) { ker=BLFEvalAssemble2D<2,3>; }
+      if (d==3 && q==4) { ker=BLFEvalAssemble2D<3,4>; }
+      if (d==4 && q==5) { ker=BLFEvalAssemble2D<4,5>; }
+      if (d==5 && q==6) { ker=BLFEvalAssemble2D<5,6>; }
+   }
+
+   if (dim==3)
+   {
+      if (d==1 && q==1) { ker=BLFEvalAssemble3D<1,1>; }
+      if (d==2 && q==2) { ker=BLFEvalAssemble3D<2,2>; }
+      if (d==3 && q==3) { ker=BLFEvalAssemble3D<3,3>; }
+      if (d==4 && q==4) { ker=BLFEvalAssemble3D<4,4>; }
+      if (d==5 && q==5) { ker=BLFEvalAssemble3D<5,5>; }
+      if (d==2 && q==3) { ker=BLFEvalAssemble3D<2,3>; }
+      if (d==3 && q==4) { ker=BLFEvalAssemble3D<3,4>; }
+      if (d==4 && q==5) { ker=BLFEvalAssemble3D<4,5>; }
+      if (d==5 && q==6) { ker=BLFEvalAssemble3D<5,6>; }
+   }
+
+   MFEM_VERIFY(ker, "No kernel ndof " << d << " nqpt " << q);
```

Comment:
Would the `BLFEvalAssemble` kernels work as fallbacks, and if they do, would it possible to add them ?

Reply:
We should add them, I can verify the domain ones.

Reply:
Sounds good, could you push that here and then I'll add the boundary?

Reply:
It turns out that that the `ker` is already initialized as `BLFEvalAssemble2D<>` or `BLFEvalAssemble3D<>`.